.github/workflows/ci.yml

```diff
@@ -32,10 +32,10 @@ jobs:
           path: |
             ~/.cache/pip
             ~/.cache/pre-commit
-          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
+          key: ${{ runner.os }}-python-${{ matrix.python-version }}-pip-${{ hashFiles('requirements.txt') }}-git-${{ github.sha }}
```

Comment:
Why adding the `-git-${{ github.sha }}` suffix?
It doesn't match the `restore-key`, so that probably won't work

Reply:
This is a bit complicated.

`actions/cache` is searching for existing cache items using the following logic:
1. Use `key`
2. If nothing found, iterate over `restore-keys` and use each item as a prefix to search for any existing cache item
3. If multiple items are found, use the most recent one

When all the steps of a job have been finished successfully, then cache will be saved:
1. If cache was fetched by `key`, it will not be updated
2. If cache was fetched by one of `restore-keys`, it'll be saved

The issue I'm trying to solve here:
1. You had some existing cache item, and it was fetched by `key`
2. You've updated `.pre-commit-config.yml`, for example some hook version was changed
3. `pre-commit` will download this new hook and store it into `~/.cache/pre-commit`
4. Because of item 1 cache will not be updated
5. Next run of this this worlflow will lead to downloading this updated hook again and again, until `key` will be updated. In this particular case this can be done by updating `dev-requirements.txt` file. Until then workflow will run slowly.

There are 2 possible ways to solve this issue:
1. Add all the files change of which ones will lead to changing the cached paths content. This could be not an easy task because there are a lot of tools with their own config files which could do that.
2. Add to the key some part which is constantly changing, like commit SHA. This will lead to updating cache on every commit.

Reply:
Thank you for the detailed explanation!
Very insightful 🙂
Merging this now